### PR TITLE
SNOW-692945: refactoring for resultset C wrapper

### PR DESCRIPTION
**Pull request by SimbaGithub**
Preparing for adding multiple statements support, refactoring on the resultset C wrapper. Remove unnecessary code and make it more cleaner for adding new feature.

### CODE REVIEW

File: cpp/lib/ResultSetArrow.cpp [hunk 1]
```diff
@@ -44,6 +43,36 @@ ResultSetArrow::ResultSetArrow(
...
 }
 
+ResultSetArrow::ResultSetArrow(
+    cJSON * jsonRowset64,
+    SF_COLUMN_DESC * metadata,
+    std::string tzString
+) :
+    ResultSet(metadata, tzString)
+{
+    m_queryResultFormat = SF_ARROW_FORMAT;
```
Comment (sfc-gh-dprzybysz) [inline cpp/lib/ResultSetArrow.cpp:53]: can we pass the format in base class constructor above?

File: cpp/lib/result_set.cpp [hunk 2]
```diff
@@ -4,382 +4,283 @@
...
     )
     {
-        switch (*query_result_format)
+        switch (query_result_format)
         {
-            case ARROW_FORMAT:
-                return rs_arrow_create_with_json_result(json_rowset, metadata, tz_string);
-            case JSON_FORMAT:
-                return rs_json_create(json_rowset, metadata, tz_string);
+#ifndef SF_WIN32
```
Comment (sfc-gh-dprzybysz) [inline cpp/lib/result_set.cpp:23]: we don't support ARROW on windows?

File: include/snowflake/client.h [hunk 3]
```diff
@@ -434,6 +434,16 @@ typedef struct SF_CHUNK_DOWNLOADER SF_CHUNK_DOWNLOADER;
...
 
+typedef void* result_set_ptr;
+
+/**
+ * An enumeration over all supported query result formats.
+ */
+typedef enum QueryResultFormat
+{
+  SF_ARROW_FORMAT, SF_JSON_FORMAT, SF_FORMAT_MAX
+} QueryResultFormat_t;
```
Comment (sfc-gh-dprzybysz) [inline include/snowflake/client.h:445]: cannot we name the type the same as the enum? without _t?

File: include/snowflake/client.h [hunk 4]
```diff
@@ -434,6 +434,16 @@ typedef struct SF_CHUNK_DOWNLOADER SF_CHUNK_DOWNLOADER;
...
 typedef struct SF_PUT_GET_RESPONSE SF_PUT_GET_RESPONSE;
 
+typedef void* result_set_ptr;
+
+/**
+ * An enumeration over all supported query result formats.
+ */
+typedef enum QueryResultFormat
+{
+  SF_ARROW_FORMAT, SF_JSON_FORMAT, SF_FORMAT_MAX
```
Comment (sfc-gh-dprzybysz) [inline include/snowflake/client.h:444]: what is `SF_FORMAT_MAX`?

File: include/snowflake/client.h [hunk 5]
```diff
@@ -443,9 +453,9 @@ typedef struct SF_STMT {
     char request_id[SF_UUID4_LEN];
     SF_ERROR_STRUCT error;
     SF_CONNECT *connection;
-    void *qrf;
+    QueryResultFormat_t qrf;
```
Comment (sfc-gh-dprzybysz) [inline include/snowflake/client.h:456]: is this change compatible for users? our public API is changing there - don't we need * anymore?

File: cpp/lib/result_set.cpp (hunk 2 above)
Comment (SimbaGithub) [reply cpp/lib/result_set.cpp:23]: That's for Windows 32-bit only. Arrow doesn't support win 32-bit build when we adding support for arrow format. Haven't tried with latest version but the arrow issue still open so unlikely it's solved.
https://github.com/apache/arrow/issues/32111

File: include/snowflake/client.h (hunk 5 above)
Comment (sfc-gh-ext-simba-hx) [reply include/snowflake/client.h:464]: it was void* and I don't think application would use it without knowing the actual data type it points to.
It was actually pointing to QueryResultFormat_t type which was dynamically allocated, having the enum data directly would make much more sense.

File: cpp/lib/ResultSetArrow.cpp (hunk 1 above)
Comment (sfc-gh-ext-simba-hx) [reply cpp/lib/ResultSetArrow.cpp]: done

File: include/snowflake/client.h (hunk 3 above)
Comment (sfc-gh-ext-simba-hx) [reply include/snowflake/client.h]: done

File: include/snowflake/client.h (hunk 4 above)
Comment (sfc-gh-ext-simba-hx) [reply include/snowflake/client.h]: renamed as `SF_FORMAT_UNKNOWN`, used as initial value.

**Review by sfc-gh-dstempniak — APPROVED** [review]

**Comment by sfc-gh-ext-simba-hx** [comment]
@sfc-gh-dstempniak Please help to check what's the failure with `continuous-integration/jenkins/pr-head`

**Comment by sfc-gh-dstempniak** [comment]
@sfc-gh-ext-simba-hx  One test failed on Macaarch64:
```
2024-09-05 00:02:15 56: [ RUN      ] test_simple_put_use_dev_urandom
2024-09-05 00:02:15 56: [  ERROR   ] --- Test failed with exception: Segmentation fault: 11(11)
2024-09-05 00:02:15 56: [  FAILED  ] test_simple_put_use_dev_urandom
```
After rerun it succeeded:
```
2024-09-05 12:21:30 56: [ RUN      ] test_simple_put_use_dev_urandom
2024-09-05 12:21:34 56: [       OK ] test_simple_put_use_dev_urandom
```

File: include/snowflake/client.h (hunk 5 above)
Comment (sfc-gh-dprzybysz) [reply include/snowflake/client.h]: And that't the reason why we cannot merge it even now - right?

**Review by sfc-gh-dprzybysz — APPROVED** [review]

File: cpp/lib/result_set.cpp [hunk 6]
```diff
@@ -4,382 +4,283 @@
...
+            case SF_ARROW_FORMAT:
+                return new Snowflake::Client::ResultSetArrow(json_rowset, metadata, std::string(tz_string));
```
Comment (sfc-gh-jszczerbinski) [inline cpp/lib/result_set.cpp:25]: Create factory method in ResultSet?

File: cpp/lib/result_set.cpp [hunk 7]
```diff
@@ -4,382 +4,283 @@
...
     )
     {
-        switch (*query_result_format)
+        switch (query_result_format)
         {
-            case ARROW_FORMAT:
-                return rs_arrow_create_with_chunk((NON_JSON_RESP*)initial_chunk, metadata, tz_string);
-            case JSON_FORMAT:
-                return rs_json_create((cJSON*)initial_chunk, metadata, tz_string);
+#ifndef SF_WIN32
```
Comment (sfc-gh-jszczerbinski) [inline cpp/lib/result_set.cpp:43]: Create factory method in ResultSet?

**Review by sfc-gh-jszczerbinski — CHANGES_REQUESTED** [review]

File: cpp/lib/result_set.cpp [hunk 8]
```diff
@@ -4,382 +4,283 @@
...
-        switch (*query_result_format){
-            case ARROW_FORMAT:
-                rs_arrow_destroy((rs_arrow_t *) rs);
+        if (!rs)
+        {
+            return;
+        }
+        QueryResultFormat query_result_format =
+            static_cast<Snowflake::Client::ResultSet*>(rs)->getResultFormat();
+        switch (query_result_format){
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-hx) [reply cpp/lib/result_set.cpp]: done

File: cpp/lib/result_set.cpp [hunk 9]
```diff
@@ -4,382 +4,283 @@
...
+    if (ptr == NULL) { return SF_STATUS_ERROR_NULL_POINTER; } \
+}
+
     SF_STATUS STDCALL
-    rs_append_chunk(void * rs, QueryResultFormat_t * query_result_format, void * chunk)
+    rs_append_chunk(result_set_ptr rs, void * chunk)
     {
-        switch (*query_result_format)
+        ERROR_IF_NULL(rs);
+        QueryResultFormat query_result_format =
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-hx) [reply cpp/lib/result_set.cpp]: done

File: cpp/lib/result_set.cpp (hunk 6 above)
Comment (sfc-gh-ext-simba-hx) [reply cpp/lib/result_set.cpp]: done

File: cpp/lib/result_set.cpp (hunk 7 above)
Comment (sfc-gh-ext-simba-hx) [reply cpp/lib/result_set.cpp]: done

**Comment by sfc-gh-ext-simba-hx** [comment]
@sfc-gh-dprzybysz 
Please help to check what's the failure with `continuous-integration/jenkins/pr-head` ?

**Comment by sfc-gh-jszczerbinski** [comment]
Failure on Linux-aarch64:
```sh
2024-10-16 18:40:55  /mnt/host/deps-build/linux/Release/arrow/lib64/libarrow.a(type.cc.o): In function `arrow::Field::MergeOptions::ToString[abi:cxx11]() const':
2024-10-16 18:40:55  type.cc:(.text+0x4d74): undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()'
2024-10-16 18:40:55  /mnt/host/deps-build/linux/Release/arrow/lib64/libarrow.a(type.cc.o): In function `arrow::Field::ToString[abi:cxx11](bool) const':
2024-10-16 18:40:55  type.cc:(.text+0x5144): undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()'
2024-10-16 18:40:55  /mnt/host/deps-build/linux/Release/arrow/lib64/libarrow.a(type.cc.o): In function `arrow::TypeHolder::ToString[abi:cxx11](std::vector<arrow::TypeHolder, std::allocator<arrow::TypeHolder> > const&)':
2024-10-16 18:40:55  type.cc:(.text+0x541c): undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()'
2024-10-16 18:40:55  /mnt/host/deps-build/linux/Release/arrow/lib64/libarrow.a(type.cc.o): In function `arrow::ListType::ToString[abi:cxx11]() const':
2024-10-16 18:40:55  type.cc:(.text+0x56d4): undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()'
2024-10-16 18:40:55  /mnt/host/deps-build/linux/Release/arrow/lib64/libarrow.a(type.cc.o): In function `arrow::LargeListType::ToString[abi:cxx11]() const':
2024-10-16 18:40:55  type.cc:(.text+0x58f4): undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()'
2024-10-16 18:40:55  /mnt/host/deps-build/linux/Release/arrow/lib64/libarrow.a(type.cc.o):type.cc:(.text+0x5b14): more undefined references to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()' follow
```

Seems like weird error, consider rebasing - maybe it will go away

**Review by sfc-gh-jszczerbinski — APPROVED** [review]

**Pull request merged**